### ISSUE TITLE
DeprecationWarnings with Python 3.4

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1409]
When warnings are enabled (using e.g. `PYTHONWARNINGS="d` environmental variable), then test suite of Sphinx triggers some DeprecationWarnings with Python 3.4.

```
$ PYTHONPATH="build/lib" PYTHONWARNINGS="d,i::ImportWarning" nosetests-3.4 -v -w tests
...
test_application.test_output ... /tmp/sphinx/build/lib/sphinx/util/pycompat.py:82: DeprecationWarning: 'U' mode is deprecated
  f = open(filepath, 'rbU')
ok
...
test_build.test_nonascii_path ... /usr/lib64/python3.4/codecs.py:896: DeprecationWarning: 'U' mode is deprecated
  file = builtins.open(filename, mode, buffering)
ok
...
```

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1409
- Originally reported by: Arfrever Frehtes Taifersar Arahesis
- Originally created at: 2014-03-03T13:36:12.387

**Comment by tk0miya** [comment]
Now, I got a following result:

```
$ git show -s
commit 7894f0bd9ca6796e329797a4a4da97dec4ff72f6
Author: Takeshi KOMIYA <i.tkomiya@gmail.com>
Date:   Sat Jan 9 19:56:07 2016 +0900

    Fix #2225: If the option does not begin with dash, linking is not performed
$ PYTHONWARNINGS="d,i::ImportWarning" tox -e py27,py35
GLOB sdist-make: /Users/tkomiya/work/sphinx/setup.py
py27 inst-nodeps: /Users/tkomiya/work/sphinx/.tox/dist/Sphinx-1.3.3.dev20160109.zip
py27 installed: -f /Users/tkomiya/.pip/wheel,alabaster==0.7.7,Babel==2.1.1,docutils==0.12,funcsigs==0.4,Jinja2==2.8,MarkupSafe==0.23,mock==1.3.0,nose==1.3.7,pbr==1.8.1,Pygments==2.0.2,pytz==2015.7,six==1.10.0,snowballstemmer==1.2.0,Sphinx==1.3.3.dev20160109,sphinx-rtd-theme==0.1.9,SQLAlchemy==1.0.11,wheel==0.24.0,Whoosh==2.7.0
py27 runtests: PYTHONHASHSEED='1903787847'
py27 runtests: commands[0] | /Users/tkomiya/work/sphinx/.tox/py27/bin/python tests/run.py
Checking dependencies...
Temporary files will be placed in /Users/tkomiya/work/sphinx/.tox/py27/testbuild.
Running Sphinx test suite (with Python 2.7.11)...
............................../Users/tkomiya/work/sphinx/sphinx/ext/todo.py:56: DeprecationWarning: make_admonition is deprecated, use docutils.parsers.rst.directives.admonitions.BaseAdmonition instead
  self.block_text, self.state, self.state_machine)
................................................................................................................................................................................................................................................................................................................................................................................................................................................./Users/tkomiya/.pyenv/versions/2.7.11/lib/python2.7/xml/etree/ElementTree.py:1627: DeprecationWarning: This method of XMLParser is deprecated.  Define doctype() method on the TreeBuilder target.
  DeprecationWarning,
................................................................................................................................................................................................................................................................S..................................................
----------------------------------------------------------------------
Ran 770 tests in 83.995s

OK (SKIP=1)
py27 runtests: commands[1] | sphinx-build -q -W -b html -d /Users/tkomiya/work/sphinx/.tox/py27/tmp/doctrees doc /Users/tkomiya/work/sphinx/.tox/py27/tmp/html
/Users/tkomiya/work/sphinx/.tox/py27/lib/python2.7/site-packages/sphinx/ext/todo.py:56: DeprecationWarning: make_admonition is deprecated, use docutils.parsers.rst.directives.admonitions.BaseAdmonition instead
  self.block_text, self.state, self.state_machine)
py35 inst-nodeps: /Users/tkomiya/work/sphinx/.tox/dist/Sphinx-1.3.3.dev20160109.zip
py35 installed: /Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/distutils/__init__.py:4: PendingDeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses,  import imp,-f /Users/tkomiya/.pip/wheel,alabaster==0.7.7,Babel==2.2.0,docutils==0.12,Jinja2==2.8,MarkupSafe==0.23,nose==1.3.7,Pygments==2.0.2,pytz==2015.7,six==1.10.0,snowballstemmer==1.2.1,Sphinx==1.3.3.dev20160109,sphinx-rtd-theme==0.1.9,SQLAlchemy==1.0.11,wheel==0.24.0,Whoosh==2.7.0
py35 runtests: PYTHONHASHSEED='1903787847'
py35 runtests: commands[0] | /Users/tkomiya/work/sphinx/.tox/py35/bin/python tests/run.py
Checking dependencies...
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/nose/importer.py:12: PendingDeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  from imp import find_module, load_module, acquire_lock, release_lock
Temporary files will be placed in /Users/tkomiya/work/sphinx/.tox/py35/testbuild.
Running Sphinx test suite (with Python 3.5.1)...
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/nose/util.py:453: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  inspect.getargspec(func)
/Users/tkomiya/work/sphinx/sphinx/util/pycompat.py:106: DeprecationWarning: 'U' mode is deprecated
  f = open(filepath, 'rbU')
./Users/tkomiya/work/sphinx/sphinx/util/pycompat.py:106: DeprecationWarning: 'U' mode is deprecated
  f = open(filepath, 'rbU')
............./Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/nose/util.py:453: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  inspect.getargspec(func)
/Users/tkomiya/work/sphinx/tests/path.py:145: DeprecationWarning: 'U' mode is deprecated
  with open(self, mode='U', encoding=encoding, **kwargs) as f:
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/docutils/io.py:236: DeprecationWarning: 'U' mode is deprecated
  self.source = open(source_path, mode, **kwargs)
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site.py:429: DeprecationWarning: 'U' mode is deprecated
  fp = open(filename, "rU")
................/Users/tkomiya/work/sphinx/sphinx/ext/todo.py:56: DeprecationWarning: make_admonition is deprecated, use docutils.parsers.rst.directives.admonitions.BaseAdmonition instead
  self.block_text, self.state, self.state_machine)
............................./Users/tkomiya/work/sphinx/sphinx/util/pycompat.py:106: DeprecationWarning: 'U' mode is deprecated
  f = open(filepath, 'rbU')
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/docutils/io.py:236: DeprecationWarning: 'U' mode is deprecated
  self.source = open(source_path, mode, **kwargs)
.............................................................................................................................................................../Users/tkomiya/work/sphinx/tests/path.py:145: DeprecationWarning: 'U' mode is deprecated
  with open(self, mode='U', encoding=encoding, **kwargs) as f:
............................................................................................................................................................................................................/Users/tkomiya/work/sphinx/sphinx/util/pycompat.py:106: DeprecationWarning: 'U' mode is deprecated
  f = open(filepath, 'rbU')
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/docutils/io.py:236: DeprecationWarning: 'U' mode is deprecated
  self.source = open(source_path, mode, **kwargs)
/Users/tkomiya/work/sphinx/tests/path.py:145: DeprecationWarning: 'U' mode is deprecated
  with open(self, mode='U', encoding=encoding, **kwargs) as f:
./Users/tkomiya/work/sphinx/sphinx/ext/todo.py:56: DeprecationWarning: make_admonition is deprecated, use docutils.parsers.rst.directives.admonitions.BaseAdmonition instead
  self.block_text, self.state, self.state_machine)
./Users/tkomiya/work/sphinx/sphinx/util/pycompat.py:106: DeprecationWarning: 'U' mode is deprecated
  f = open(filepath, 'rbU')
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/docutils/io.py:236: DeprecationWarning: 'U' mode is deprecated
  self.source = open(source_path, mode, **kwargs)
./Users/tkomiya/work/sphinx/tests/path.py:145: DeprecationWarning: 'U' mode is deprecated
  with open(self, mode='U', encoding=encoding, **kwargs) as f:
............./Users/tkomiya/work/sphinx/sphinx/util/pycompat.py:106: DeprecationWarning: 'U' mode is deprecated
  f = open(filepath, 'rbU')
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/docutils/io.py:236: DeprecationWarning: 'U' mode is deprecated
  self.source = open(source_path, mode, **kwargs)
/Users/tkomiya/work/sphinx/tests/path.py:145: DeprecationWarning: 'U' mode is deprecated
  with open(self, mode='U', encoding=encoding, **kwargs) as f:
....................................................../Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/nose/util.py:453: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  inspect.getargspec(func)
/Users/tkomiya/work/sphinx/sphinx/ext/todo.py:56: DeprecationWarning: make_admonition is deprecated, use docutils.parsers.rst.directives.admonitions.BaseAdmonition instead
  self.block_text, self.state, self.state_machine)
........./Users/tkomiya/work/sphinx/tests/path.py:145: DeprecationWarning: 'U' mode is deprecated
  with open(self, mode='U', encoding=encoding, **kwargs) as f:
./Users/tkomiya/work/sphinx/sphinx/util/pycompat.py:106: DeprecationWarning: 'U' mode is deprecated
  f = open(filepath, 'rbU')
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/docutils/io.py:236: DeprecationWarning: 'U' mode is deprecated
  self.source = open(source_path, mode, **kwargs)
...../Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/nose/util.py:453: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  inspect.getargspec(func)
........................................................../Users/tkomiya/work/sphinx/sphinx/ext/todo.py:56: DeprecationWarning: make_admonition is deprecated, use docutils.parsers.rst.directives.admonitions.BaseAdmonition instead
  self.block_text, self.state, self.state_machine)
./Users/tkomiya/work/sphinx/sphinx/util/pycompat.py:106: DeprecationWarning: 'U' mode is deprecated
  f = open(filepath, 'rbU')
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/docutils/io.py:236: DeprecationWarning: 'U' mode is deprecated
  self.source = open(source_path, mode, **kwargs)
/Users/tkomiya/work/sphinx/tests/path.py:145: DeprecationWarning: 'U' mode is deprecated
  with open(self, mode='U', encoding=encoding, **kwargs) as f:
................/Users/tkomiya/work/sphinx/tests/test_intl.py:313: ResourceWarning: unclosed file <_io.TextIOWrapper name=path('/Users/tkomiya/work/sphinx/.tox/py35/testbuild/test-intl/definition_terms.po') mode='r' encoding='UTF-8'>
  expect = pofile.read_po((app.srcdir / 'definition_terms.po').open())
/Users/tkomiya/work/sphinx/tests/test_intl.py:314: ResourceWarning: unclosed file <_io.TextIOWrapper name=path('/Users/tkomiya/work/sphinx/.tox/py35/testbuild/test-intl/_build/gettext/definition_terms.pot') mode='r' encoding='UTF-8'>
  actual = pofile.read_po((app.outdir / 'definition_terms.pot').open())
......../Users/tkomiya/work/sphinx/tests/test_intl.py:319: ResourceWarning: unclosed file <_io.TextIOWrapper name=path('/Users/tkomiya/work/sphinx/.tox/py35/testbuild/test-intl/glossary_terms.po') mode='r' encoding='UTF-8'>
  expect = pofile.read_po((app.srcdir / 'glossary_terms.po').open())
/Users/tkomiya/work/sphinx/tests/test_intl.py:320: ResourceWarning: unclosed file <_io.TextIOWrapper name=path('/Users/tkomiya/work/sphinx/.tox/py35/testbuild/test-intl/_build/gettext/glossary_terms.pot') mode='r' encoding='UTF-8'>
  actual = pofile.read_po((app.outdir / 'glossary_terms.pot').open())
......./Users/tkomiya/work/sphinx/tests/test_intl.py:327: ResourceWarning: unclosed file <_io.TextIOWrapper name=path('/Users/tkomiya/work/sphinx/.tox/py35/testbuild/test-intl/glossary_terms_inconsistency.po') mode='r' encoding='UTF-8'>
  expect = pofile.read_po((app.srcdir / 'glossary_terms_inconsistency.po').open())
/Users/tkomiya/work/sphinx/tests/test_intl.py:328: ResourceWarning: unclosed file <_io.TextIOWrapper name=path('/Users/tkomiya/work/sphinx/.tox/py35/testbuild/test-intl/_build/gettext/glossary_terms_inconsistency.pot') mode='r' encoding='UTF-8'>
  actual = pofile.read_po((app.outdir / 'glossary_terms_inconsistency.pot').open())
.................................................................................................../Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/nose/util.py:453: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  inspect.getargspec(func)
......................S/Users/tkomiya/work/sphinx/sphinx/ext/todo.py:56: DeprecationWarning: make_admonition is deprecated, use docutils.parsers.rst.directives.admonitions.BaseAdmonition instead
  self.block_text, self.state, self.state_machine)
.............../Users/tkomiya/work/sphinx/tests/test_util_i18n.py:58: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/tkomiya/work/sphinx/.tox/py35/testbuild/tmp0y_f5a6n/test.mo'>
  assert read_mo(open(cat.mo_path, 'rb')) is not None
...................................
----------------------------------------------------------------------
Ran 769 tests in 71.998s

OK (SKIP=1)
py35 runtests: commands[1] | sphinx-build -q -W -b html -d /Users/tkomiya/work/sphinx/.tox/py35/tmp/doctrees doc /Users/tkomiya/work/sphinx/.tox/py35/tmp/html
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/sphinx/util/pycompat.py:106: DeprecationWarning: 'U' mode is deprecated
  f = open(filepath, 'rbU')
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/docutils/io.py:236: DeprecationWarning: 'U' mode is deprecated
  self.source = open(source_path, mode, **kwargs)
/Users/tkomiya/work/sphinx/.tox/py35/lib/python3.5/site-packages/sphinx/ext/todo.py:56: DeprecationWarning: make_admonition is deprecated, use docutils.parsers.rst.directives.admonitions.BaseAdmonition instead
  self.block_text, self.state, self.state_machine)
______________________________________________________________ summary _______________________________________________________________
  py27: commands succeeded
  py35: commands succeeded
  congratulations :)
```

**Comment by tk0miya** [comment]
Refactored at 053d780 (stable) and 1e2fbd463b2b2d22f0ed866ec512968e39af80a5 (master).
After that, no warnings by following command:

```
PYTHONWARNINGS="d,i::ImportWarning,i:::sphinx.util.pycompat,i:::path,i:::docutils.io,i:::nose.util" tox -e py27,py35
```

I ignored some warnings by following reason:
- `sphinx.util.pycompat`: This module uses deprecated features with the intention to keep compatibility between multiple python versions
- `path`: This module is used to testcases (not features)
- `docutils.io`, `nose.util`: These are not parts of Sphinx
